include/SDL3/SDL_ioasync.h

```diff
@@ -0,0 +1,500 @@
+ * might take longer than the life of the calling function to complete!
+ *
+ * An SDL_IOAsyncGroup must be specified. The newly-created SDL_IOAsyncTask
+ * will be added to it. A group is used to query multiple async tasks at once.
+ *
+ * \param ioasync a pointer to an SDL_IOAsync structure.
+ * \param ptr a pointer to a buffer to read data into.
+ * \param offset the position to start reading in the data source.
+ * \param size the number of bytes to read from the data source.
+ * \param group a group to add the new SDL_IOAsync to.
+ * \param userdata an app-defined pointer that will be provided with the task results.
+ * \returns A new task handle if a task was started, NULL on complete failure;
+ *          call SDL_GetError() for more information.
+ *
+ * \threadsafety It is safe to call this function from any thread.
+ *
+ * \since This function is available since SDL 3.0.0.
+ *
+ * \sa SDL_WriteIOAsync
+ * \sa SDL_CreateIOAsyncGroup
+ * \sa SDL_GetIOAsyncTaskResult
+ */
+extern SDL_DECLSPEC SDL_IOAsyncTask * SDLCALL SDL_ReadIOAsync(SDL_IOAsync *ioasync, void *ptr, Uint64 offset, Uint64 size, SDL_IOAsyncGroup *group, void *userdata);
+
+/**
+ * Start an async write.
+ *
+ * This function writes `size` bytes from `offset` position in the data
+ * source to the area pointed at by `ptr`.
+ *
+ * This function returns as quickly as possible; it does not wait for the
+ * write to complete. On a successful return, this work will continue in the
+ * background. If the work begins, even failure is asynchronous: a failing
+ * return value from this function only means the work couldn't start at all.
+ *
+ * `ptr` must remain available until the work is done, and may be accessed by
+ * the system at any time until then. Do not allocate it on the stack, as this
+ * might take longer than the life of the calling function to complete!
+ *
+ * An SDL_IOAsyncGroup must be specified. The newly-created SDL_IOAsyncTask
+ * will be added to it. A group is used to query multiple async tasks at once.
+ *
+ * \param ioasync a pointer to an SDL_IOAsync structure.
+ * \param ptr a pointer to a buffer to write data from.
+ * \param offset the position to start writing to the data source.
+ * \param size the number of bytes to write to the data source.
+ * \param group a group to add the new SDL_IOAsync to.
+ * \param userdata an app-defined pointer that will be provided with the task results.
+ * \returns A new task handle if a task was started, NULL on complete failure;
+ *          call SDL_GetError() for more information.
+ *
+ * \threadsafety It is safe to call this function from any thread.
+ *
+ * \since This function is available since SDL 3.0.0.
+ *
+ * \sa SDL_ReadIOAsync
+ * \sa SDL_CreateIOAsyncGroup
+ * \sa SDL_GetIOAsyncTaskResult
+
+ */
+extern SDL_DECLSPEC SDL_IOAsyncTask * SDLCALL SDL_WriteIOAsync(SDL_IOAsync *ioasync, void *ptr, Uint64 offset, Uint64 size, SDL_IOAsyncGroup *group, void *userdata);
```

Comment:
It seems really odd that each operation could be part of a different group. Do async I/O APIs really work this way?

Reply:
It's not impossible you'd want to have a single packfile with game data, and put all your images in one group and audio in another, and have different code process each, knowing the group only has one type of data or another when it goes to pull from it.

(But I'll defer to @jaedan to say if that's something no one should ever actually do.)

Reply:
I was more thinking about native async I/O APIs, which I haven't looked at, and I should before questioning your design. :)

Reply:
Basically Windows IoRing and Linux io_uring work like more complex versions of this: "take this file handle, do this thing to it, tell me when you're done." On Linux, the operations are literally just a list of syscalls: read, write, etc...but it also works with things like socket connect(), etc. It's kinda wild.

Older APIs _don't_ work like this, but we're just using the generic threadpool for those and I don't expect that to change.

Reply:
[Linux aio also has big caveats and limitations that io_uring doesn't.](https://stackoverflow.com/questions/34572559/asynchronous-io-io-submit-latency-in-ubuntu-linux). It works extremely well for specific I/O patterns but it would require some significant shimming to be generally useful, and after that I'm not convinced it would actually outperform a threadpool using normal buffered pread/pwrite.

Reply:
> I was more thinking about native async I/O APIs, which I haven't looked at, and I should before questioning your design. :)

the only significant native async API for regular files that requires associating a file with a single context is Windows' IOCP, but it's fairly typical to use a single completion port for all I/O in a process anyway.